Add nolearn package

jfy133 opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/main/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found some lint.

Here's what I've got...


For **recipes/nolearn**:

 * noarch: python recipes are required to have a lower bound on the python version. Typically this means putting `python >=3.6` in **both** `host` and `run` but you should check upstream for the package's Python compatibility.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found some lint.

Here's what I've got...


For **recipes/nolearn**:

 * ``requirements: host: python>=3.4`` must contain a space between the name and the pin, i.e. ``python >=3.4``
 * ``requirements: run: python>=3.6`` must contain a space between the name and the pin, i.e. ``python >=3.6``
 * noarch: python recipes are required to have a lower bound on the python version. Typically this means putting `python >=3.6` in **both** `host` and `run` but you should check upstream for the package's Python compatibility.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found it was in an excellent condition.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found some lint.

Here's what I've got...


For **recipes/nolearn**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found some lint.

Here's what I've got...


For **recipes/nolearn**:

 * `noarch` packages can't have selectors. If the selectors are necessary, please remove `noarch: python`.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found it was in an excellent condition.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found some lint.

Here's what I've got...


For **recipes/nolearn**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [15]

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/nolearn```) and found it was in an excellent condition.

jfy133 commented:
@conda-forge/help-python [warning: this is my first attempt at a conda-forge recipe]

I've got the recipe to work for Linux/OSX, but I didn't understand how to get it to work with Windows due to a vc9* error. 

Note it's an old package and I'm only adding it because another (now undeveloped) tool that I need to add to bioconda for a colleague, requires it. Thus the strict pinning.

jfy133 commented:
@conda-forge-admin, please ping team

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-webservice.

I was asked to ping @conda-forge/staged-recipes and so here I am doing that.

dopplershift commented on recipes/nolearn/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+  host:
+    - pip
+    - python ==2.7.*
+    - {{ compiler('cxx') }}
+  run:
+    - {{ compiler('cxx') }}
```
It doesn't look like this package uses a C++ compiler?
```suggestion
  run:
```

dopplershift commented on recipes/nolearn/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+requirements:
+  host:
+    - pip
+    - python ==2.7.*
```
It looks like the upstream package supports Python 3 just fine. We generally aren't building for Python 2 any more. Is there any reason you're locking this to Python 2 rather than say:
```suggestion
    - python >=3.6
```

dopplershift commented on recipes/nolearn/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+    - joblib
+    - lasagne
+    - python ==2.7.*
+    - scikit-learn ==0.17*
+    - tabulate ==0.7*
```
Upstream doesn't have these version requirements, is there any reason you're locking these down?

dopplershift commented on recipes/nolearn/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+build:
+  number: 0
+  script: {{ PYTHON }} -m pip install . -vv
+  skip: True  # [win]
```
This looks like it's pure Python and could be noarch Python (which makes things much easier on the package build end):
```suggestion
  noarch: python
```
If Windows builds fail due to missing dependencies, we can just ignore the CI failure.

jfy133 commented:
Hi @dopplershift , thanks for the review and comments!

I actually went through exactly the same thought process as you (as in, everything you suggested is what I originally had the in the recipe), however each thing didn't work.

Unfortunately, I just tried looking at the old azure pipeline tests to find the errors, but it seems the logs aren't accessible anymore.

1. For `cxx` compiler I added because there was a specific error saying it could not be found during build, and when I added it started working
2. True, I will try again with 3.6 (could've been a different related error)
3. I was having version incompatiblilty issues, meaning with the error saying something like `this function has been moved into this other package`, so I thought it was safer to define all requirements as per [here](https://github.com/dnouri/nolearn/blob/master/requirements.txt)
4. I removed the `noarch` specification as per the linting comment here: `https://github.com/conda-forge/staged-recipes/pull/16860#issuecomment-965085431`

However, as I can't access the logs anymore, and as you can see from the commit history I went through trial and error, I will make a new commit with all the changes you request, so you can see the issues/errors first, and then if there is no better solution we can revert back to the current commit. I hope that's OK.

isuruf commented on recipes/nolearn/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+  host:
+    - pip
+    - python >=3.6
+    - {{ compiler('cxx') }}
```
```suggestion
```

jfy133 replied:
oops good catch!

dopplershift commented on recipes/nolearn/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+    - joblib
+    - lasagne
+    - python >=3.6
+    - scikit-learn
```
So it looks like `scikit.cross_validation` was renamed, so should be able to get away with just a bound here:
```suggestion
    - scikit-learn <0.20
```

jfy133 commented:
Ok Linux and osx passed, but windows failed with: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=409592&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b&l=823

Is that something we can 'ignore' in this case (for future reference)?

dopplershift reviewed (approved):
Ignoring windows failure for noarch due to missing dependency.

dopplershift commented:
@jfy133 Yes, for future reference, if a package is noarch and fails one of the platforms (besides linux) due to a missing dep, we just merge over the failure. Noarch packages are only built on Linux, and the dependency failure just means users can't install the package on their platform.

jfy133 commented:
> @jfy133 Yes, for future reference, if a package is noarch and fails one of the platforms (besides linux) due to a missing dep, we just merge over the failure. Noarch packages are only built on Linux, and the dependency failure just means users can't install the package on their platform.

Awesome, thanks for the explanation. Thanks for all the help!

jfy133 commented:
Hi @dopplershift - sorry to resurrect an old thread. I now remember why I originally specified python 2.7.

I added `nolearn` it for the purpose of creating a bioconda recipe for: https://bitbucket.org/gusphdproj/deeparg-ss/src/master/, 

where `deepARG` specifically was written in python 2.7. Is it possible to re-build but with a lower version of python (i.e. build with a more 'relaxed' specification)?
